### PR TITLE
add BFloat16 sparse operators on CPU: sparse_mask, add_out

### PR DESCRIPTION
**Pull request by jiayisunx**
Fixes #ISSUE_NUMBER

### CODE REVIEW

**Comment by jiayisunx** [comment]
@cpuhrsch, this PR(https://github.com/pytorch/pytorch/pull/72846) had been reverted because of the CI failures, so I fixed the failing UT and reopened this PR, please help to reviewed it again, thanks.

**Comment by cpuhrsch** [comment]
@jiayisunx - What was the error and how did you fix it?

**Comment by jiayisunx** [comment]
> @jiayisunx - What was the error and how did you fix it?

Tests fail on master due to this UT, and I changed the error requirement for BF16 datatype in this UT.

❌ Failure: TestSparseCPU.test_sparse_addmm_cpu_bfloat16
Traceback (most recent call last):
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line [37](https://github.com/pytorch/pytorch/runs/5256388151?check_suite_focus=true#step:18:37)6, in instantiated_test
    result = test(self, **param_kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 2951, in wrapped
    f(self, *args, **kwargs, coalesced=False)
  File "test_sparse.py", line 1275, in test_sparse_addmm
    test_shape(7, 8, 9, 20, False, (1, 1))
  File "test_sparse.py", line 1264, in test_shape
    self.assertEqual(Y, Y_dense)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 2112, in assertEqual
    assert_equal(
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_comparison.py", line 1064, in assert_equal
    raise error_metas[0].to_error()
AssertionError: Tensor-likes are not close!

Mismatched elements: 1 / 72 (1.4%)
Greatest absolute difference: 0.1484375 at index (1, 3) (up to 0.1 allowed)
Greatest relative difference: 0.0913461[53](https://github.com/pytorch/pytorch/runs/5256388151?check_suite_focus=true#step:18:53)84615384 at index (1, 3) (up to 0.016 allowed)

**Comment by cpuhrsch** [comment]
@jiayisunx - that's a pretty big difference, especially the relative tolerance. Is this expected?

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 1]
```diff
@@ -899,7 +899,7 @@ Tensor& s_addmm_out_sparse_dense_cpu(
   Tensor indices = sparse_._indices();
   Tensor values      = sparse_._values();
 
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(kBFloat16,
```
Comment (ngimel) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:902]: I don't think we should be enabling bfloat16 like this, in regular bfloat16 matrix multiplies intermediate accumulation is done in fp32, doing intermediate accumulation in bfloat16 (as it seems is done here) is very inaccurate.

Comment (jiayisunx) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: I see.. thanks!

**Pull request closed without merging**